### PR TITLE
[Do not merge]Harris update

### PR DESCRIPTION
**Pull request by ducha-aiki**
Added option for using spatial gradient w/o smoothing if run on spatial pyramid  (https://github.com/arraiyopensource/kornia/pull/198)

Also adds function to corrent responce by spatial pyramid sigmas for proper scale detection (similar to https://github.com/perdoch/hesaff/blob/master/pyramid.cpp#L101 )

### CODE REVIEW

File: kornia/feature/harris.py [hunk 1]
```diff
@@ -74,42 +85,70 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
         # normalize and return
-        scores_max: torch.Tensor = F.adaptive_max_pool2d(scores, output_size=1)
-        return scores / scores_max
+        if self.normalize:
+            scores_max: torch.Tensor = F.adaptive_max_pool2d(scores,
+                                                             output_size=1)
+            scores = scores / scores_max.clamp(1e-6)
+        return scores
+
+    def update_response(self, resp, sigmas):
```
Comment (edgarriba) [inline kornia/feature/harris.py]: I would put it as a separated function since it doesn't use any state of sobel. Actually sobel doesn't hold state. In that case we should consider writing it as a free function.

File: kornia/feature/harris.py [hunk 2]
```diff
@@ -29,7 +29,11 @@ class CornerHarris(nn.Module):
     :math:`k ∈ [ 0.04 , 0.06 ]`
 
     Args:
-        k (torch.Tensor): the Harris detector free parameter.
+        - k (torch.Tensor): the Harris detector free parameter.
+        - do_blur (bool): perform Gaussian smoothing.
+          Set to True, if use alone, False if on scale pyramid
+        - nms (bool): perform hard non maxima supression
+        - normalize (bool): if True, responce map is divided by max value
```
Comment (edgarriba) [inline kornia/feature/harris.py]: not convinced about the parameters names. Suggest and open to discussion:
`k` -> `factor` (or just `k_factor`)
`do_blur` -> `blur`
`nms` -> `non_maxima_supression`

Comment (edgarriba) [reply kornia/feature/harris.py]: for blur, see below

File: kornia/feature/harris.py [hunk 3]
```diff
@@ -59,11 +63,18 @@ class CornerHarris(nn.Module):
...
+    def __init__(self,
+                 k: torch.Tensor,
+                 do_blur: bool = True,
+                 nms: bool = True,
+                 normalize: bool = True) -> None:
         super(CornerHarris, self).__init__()
         self.k: torch.Tensor = k
         # TODO: add as signature parameter
         self.kernel_size: Tuple[int, int] = (3, 3)
+        self.nms = nms
```
Comment (edgarriba) [inline kornia/feature/harris.py]: `self.nms: bool`, same in below

File: kornia/feature/harris.py [hunk 4]
```diff
@@ -74,42 +85,70 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
                              .format(input.shape))
         # compute the first order gradients with sobel operator
         # TODO: implement support for kernel different than three
-        gradients: torch.Tensor = spatial_gradient(input)
+        grad_mode = 'Sobel' if self.do_blur else 'NoBlur'
```
Comment (edgarriba) [inline kornia/feature/harris.py]: we could just replace `do_blur` by `grad_mode` and pass `sobel` or `no_blur`

Comment (edgarriba) [reply kornia/feature/harris.py]: `grad_mode: str`

File: kornia/feature/harris.py [hunk 5]
```diff
@@ -74,42 +85,70 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
         trace_m: torch.Tensor = dx2 + dy2
 
         # compute the response map
-        scores: torch.Tensor = det_m - self.k * trace_m ** 2
+        scores: torch.Tensor = det_m - self.k * (trace_m ** 2)
 
         # threshold
         # TODO: add as signature parameter ?
-        scores = torch.clamp(scores, min=1e-6)
+        scores = torch.clamp(scores, min=0)
```
Comment (edgarriba) [inline kornia/feature/harris.py]: we you should make sure that absolute zero won't return nans during backprop (it sometimes happened to me)

File: kornia/feature/harris.py [hunk 6]
```diff
@@ -74,42 +85,70 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
 
         # apply non maxima suppresion
-        scores = non_maxima_suppression2d(scores, kernel_size=(3, 3))
+        if self.nms:
+            scores = non_maxima_suppression2d(scores, kernel_size=(3, 3))
```
Comment (edgarriba) [inline kornia/feature/harris.py]: should we pass kernel_size as parameter?

File: kornia/feature/harris.py [hunk 7]
```diff
@@ -74,42 +85,70 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
 
         # normalize and return
-        scores_max: torch.Tensor = F.adaptive_max_pool2d(scores, output_size=1)
-        return scores / scores_max
+        if self.normalize:
+            scores_max: torch.Tensor = F.adaptive_max_pool2d(scores,
+                                                             output_size=1)
+            scores = scores / scores_max.clamp(1e-6)
```
Comment (edgarriba) [inline kornia/feature/harris.py]: 1) be explicit in what we clamp: `scores = scores / torch.clamp(scores_max, min=1e-6)`: I tend to use torch.clamp, torch.squeeze instead of using built-in tensor functions. Just to be consistent with the existing code.
2) should 1e-6 be an eps parameter ?

File: kornia/feature/harris.py [hunk 8]
```diff
@@ -74,42 +85,70 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
...
+        if not len(resp.shape) == 4:
+            raise ValueError("Invalid input shape, we expect BxCxHxW. Got: {}"
+                             .format(input.shape))
+        if not torch.is_tensor(sigmas):
+            raise TypeError("Sigmas type is not a torch.Tensor. Got {}"
+                            .format(type(sigmas)))
+        if not len(sigmas.shape) == 2:
+            raise ValueError("Invalid sigmas shape, we expect BxC. Got: {}"
+                             .format(sigmas.shape))
+        B, CH, H, W = resp.size()
```
Comment (edgarriba) [inline kornia/feature/harris.py]: `batch_size, channels, height, width`; like to give explicit names to variables whenever I can. Very up to you that one

File: kornia/feature/harris.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/feature/harris.py]: But it depends on being "harris-related". For other detector it would not be pow(4), etc. Having free function is fine, but module should have it as well, IMO

File: kornia/feature/harris.py (hunk 2 above)
Comment (ducha-aiki) [reply kornia/feature/harris.py]: Regarding nms - it is common abbreviation.
Regarding blur - it would be not clear then, what does it mean - if the blur is already there or if it should be added 
Regarding k - I don`t have any preference and actually it was you introducing the "k" :)

File: kornia/feature/harris.py (hunk 7 above)
Comment (ducha-aiki) [reply kornia/feature/harris.py]: I strongly disagree about 1) 
For consistency we then should use it verywhere. 
Imagine smth like. torch.sum(torch.pow(x,2), dim = 1) instead of simple x.pow(2).sum(dim=1)

File: kornia/feature/harris.py (hunk 6 above)
Comment (ducha-aiki) [reply kornia/feature/harris.py]: For classical features, there is no such thing as non 3x3  nms, so I think, we should go for the constant. 
For the differentiable softnms - yes, but it will be in different PR

File: kornia/feature/harris.py (hunk 7 above)
Comment (edgarriba) [reply kornia/feature/harris.py]: mmm, okay. Like it more in my way but it's fine

File: kornia/feature/harris.py (hunk 1 above)
Comment (edgarriba) [reply kornia/feature/harris.py]: We can have a mode parameter with different implementations. The module should call the free function similar to other functions we have

Comment (edgarriba) [reply kornia/feature/harris.py]: As for the function name in case is standalone wouldn't be more appropriate something like refine or scale scores_map ? Is there any other library implementing it?

File: kornia/feature/harris.py (hunk 2 above)
Comment (edgarriba) [reply kornia/feature/harris.py]: maybe apply_nms ? I took "k" for the formulation but is zero intuitive :D
We agree with blur -> grads_mode or simply mode?

File: kornia/feature/harris.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/feature/harris.py]: rescale_scores_map is  good. Other libraries put sigma directly into to main function, which I find somewhat artificial.

Comment (ducha-aiki) [reply kornia/feature/harris.py]: > We can have a mode parameter with different implementations. The module should call the free function similar to other functions we have

Agree. I`ll refactor this. It goes to be cleaner with your suggestion
